### ISSUE TITLE
[BUG] Ignoring blank lines and comments kills multiple consecutive spaces

### ISSUE THREAD

**Issue by michael-o** [BastilleBSD/bastille#400]
**[MANDATORY] Describe the bug [MANDATORY]**
If your Bastillefile contains more than one space in one line this block:
https://github.com/BastilleBSD/bastille/blob/b85d6347dee44c943819c2d7b4eee9c5a30535b5/usr/local/share/bastille/template.sh#L267-L268
kills it.

**[MANDATORY] Bastille and FreeBSD version (paste ``bastille -v && freebsd-version -kru`` output)**
0.8.20210115
12.2-STABLE
12.2-STABLE
12.2-STABLE

**[MANDATORY] How did you install bastille? (port/pkg/git)**
ports 

**[optional] Steps to reproduce?**
Add this to your `Bastillefile`:
```
sed -i "" -e 's/# Host \*/Host \*/' -e 's/#   GSSAPIAuthentication no/  GSSAPIAuthentication yes/' /etc/ssh/ssh_config
```

Run the following:
```
# grep -v '^[[:blank:]]*$' "/usr/local/bastille/templates/lda/base/Bastillefile" | grep -v '^[[:blank:]]*#'
ARG HOST_RESOLV_CONF=/etc/resolv.conf
ARG DOMAIN_NAME=ad001.siemens.net
CMD touch /etc/rc.conf
SYSRC syslogd_flags="-N -ss"
SYSRC sendmail_enable="NO"
SYSRC clear_tmp_enable="YES"
SYSRC cron_flags="-J 60"
CP "${HOST_RESOLV_CONF}" etc/resolv.conf
OVERLAY etc
RENDER etc/hosts
CMD sed -i "" -e 's/#PermitRootLogin no/PermitRootLogin yes/' -e 's/#GSSAPIAuthentication no/GSSAPIAuthentication yes/'  /etc/ssh/sshd_config
CMD sed -i "" -e 's/# Host \*/Host \*/' -e 's/#   GSSAPIAuthentication no/  GSSAPIAuthentication yes/' /etc/ssh/ssh_config
SYSRC sshd_enable="YES"
```
Looks good. Now do this:
```
# SCIRPT="$(grep -v '^[[:blank:]]*$' "/usr/local/bastille/templates/lda/base/Bastillefile" | grep -v '^[[:blank:]]*#')"
# echo $SCIRPT
ARG HOST_RESOLV_CONF=/etc/resolv.conf ARG DOMAIN_NAME=ad001.siemens.net CMD touch /etc/rc.conf SYSRC syslogd_flags="-N -ss" SYSRC sendmail_enable="NO" SYSRC clear_tmp_enable="YES" SYSRC cron_flags="-J 60" CP "${HOST_RESOLV_CONF}" etc/resolv.conf OVERLAY etc RENDER etc/hosts CMD sed -i "" -e 's/#PermitRootLogin no/PermitRootLogin yes/' -e 's/#GSSAPIAuthentication no/GSSAPIAuthentication yes/' /etc/ssh/sshd_config CMD sed -i "" -e 's/# Host \*/Host \*/' -e 's/# GSSAPIAuthentication no/ GSSAPIAuthentication yes/' /etc/ssh/ssh_config SYSRC sshd_enable="YES"
```

My spaces are now gone and `ssh_config` cannot be modified by `sed`.

**[optional] Expected behavior**
No mangling is multiple consecutive spaces

**[optional]  Additional context**
I guess a while loop would be much better here.
A workaround is to write instead of a space a `\ `.

**Comment by bmac2** [comment]
@michael-o  so if I am understanding this, it is the sed statement that is messing up the spaces?  Only that line???  Your last statement makes me ask is this an issue with how that sed line is escaped for spaces and NOT an actual Bastillefile error.  Am I reading this wrong???

**Comment by michael-o** [comment]
> @michael-o so if I am understanding this, it is the sed statement that is messing up the spaces? Only that line??? Your last statement makes me ask is this an issue with how that sed line is escaped for spaces and NOT an actual Bastillefile error. Am I reading this wrong???

IIRC, the sed statement isn't a problem, but the preprocessing shreds it, not putting in-line spaces into account. The combination of `SCRIPT` and the subshell.

**Comment by tschettervictor** [comment]
This might be fixed with PR #761 
I believe the problem is that it isn't quoting the entire command.

Can you check? @michael-o

**Comment by michael-o** [comment]
> This might be fixed with PR #761 I believe the problem is that it isn't quoting the entire command.
> 
> Can you check? @michael-o

Will check next year as soon as I get back to work.

**Comment by michael-o** [comment]
> This might be fixed with PR #761 I believe the problem is that it isn't quoting the entire command.
> 
> Can you check? @michael-o

I have applied the fix locally. With the Bastillefile as-is (with backslashes):
```
eval: 1: Syntax error: Unterminated quoted string
eval: 1: Syntax error: Error in command substitution
```
with backlashes removed:
```
[deblndw013x4j]:
sh: Syntax error: Unterminated quoted string
[deblndw013x4j]: 2

Failed to execute command: cmd
Failed to execute command: template
```

**Comment by tschettervictor** [comment]
Can you try to use the latest "template.sh" from the GitHub? It got updated.

Also, I don't think quotes are allowed in the bastillefile...

**Comment by tschettervictor** [comment]
Also, can you post your template so I can go ahead and give it a whirl on my test machine?

**Comment by michael-o** [comment]
> Also, can you post your template so I can go ahead and give it a whirl on my test machine?

My whole template:
```
CMD echo "Configuring OpenSSH"
CMD patch -V none -s -p0 < /files/patch-etc_ssh_sshd__config
# See https://github.com/BastilleBSD/bastille/issues/400
CMD sed -i "" -e 's/# Host \*/Host \*/' -e 's/#\ \ \ GSSAPIAuthentication no/\ \ GSSAPIAuthentication yes/' /etc/ssh/ssh_config
SYSRC sshd_enable="YES"
```
where the `patch` is irrelevant to the topic, you can remove it. If Bastille would work properly the backslashes would be redudant.

**Comment by michael-o** [comment]
> Can you try to use the latest "template.sh" from the GitHub? It got updated.
> 
> Also, I don't think quotes are allowed in the bastillefile...

Will do.

**Comment by michael-o** [comment]
> Can you try to use the latest "template.sh" from the GitHub? It got updated.
> 
> Also, I don't think quotes are allowed in the bastillefile...

This has introduced a regression now. `fetch`ed the file and ran my template:
```
[deblndw013x4j]:
sh: Syntax error: Unterminated quoted string
[deblndw013x4j]: 2

Failed to execute command: cmd
Failed to execute command: template
```
Master is now broken for me now.

**Comment by tschettervictor** [comment]
Remove the quotes from the first CMD.

This is a weird one. I can't figure it out. Just tried running it on mine and getting the same error.

**Comment by tschettervictor** [comment]
Ok I was able to get it to run. I'm not sure if it applied properly though. 
Can you try this?

On line 302 change 
                    _args="sh -c \"${_args}\""

to 

                    _args="sh -c '${_args}'"

And then run the template again.

**Comment by michael-o** [comment]
> Ok I was able to get it to run. I'm not sure if it applied properly though. Can you try this?
> 
> On line 302 change ` _args="sh -c \"${_args}\""`
> 
> to
> 
> ` _args="sh -c '${_args}'"`
> 
> And then run the template again.

This restores the code to the latest release and this one works, but this issue remains of course.

**Comment by tschettervictor** [comment]
Ok I got it. It's actually line 284 that is the issue. Can you try changing it from 


`            _args=$(echo "${_line}" | '{$1=""; sub(/^ */, ""); print;}' | eval "sed ${ARG_REPLACEMENTS}")
`
to
`            _args=$(echo "${_line}" | sed 's/^[^ ]*//' | eval "sed ${ARG_REPLACEMENTS}")
`
Basically the replacement "sed" command strips off the first argument from the line, which is always CMD,MOUNT, or whatever. That is all the awk command was trying to do, but apparently not correctly.

**Comment by tschettervictor** [comment]
#785

**Comment by michael-o** [comment]
> Ok I got it. It's actually line 284 that is the issue. Can you try changing it from
> 
> ` _args=$(echo "${_line}" | awk '{$1=""; sub(/^ */, ""); print;}' | eval "sed ${ARG_REPLACEMENTS}")` to ` _args=$(echo "${_line}" | sed 's/^[^ ]*//' | eval "sed ${ARG_REPLACEMENTS}")` Basically the replacement "sed" command strips off the first argument from the line, which is always CMD,MOUNT, or whatever. That is all the awk command was trying to do, but apparently not correctly.

Not working, when applied as a oneoff patch to the latest release it produces:
```
# bastille create -TV deblndw013x4j 13.3-RELEASE 10.64.105.180 bce1
Valid: (10.64.105.180).
Valid: (bce1).

Creating a thickjail. This may take a while...

pfctl: /dev/pf: No such file or directory
[deblndw013x4j]:
e0a_bastille0
e0b_bastille0
deblndw013x4j: created

[deblndw013x4j]:
Applying template: dw/vnet...
Template name/URL not recognized.
Failed to execute command: template
pfctl: /dev/pf: No such file or directory
rdr-anchor not found in pf.conf
```

So it does break applying the template in general!

**Comment by tschettervictor** [comment]
This is unrelated looks like.

Can you make sure you have the template.sh file from the current release? Ie not the one from HEAD.

**Comment by michael-o** [comment]
> This is unrelated looks like.
> 
> Can you make sure you have the template.sh file from the current release? Ie not the one from HEAD.

Yes, I do, I have reinstalled from package to make sure.

**Comment by tschettervictor** [comment]
Let me try again...

**Comment by tschettervictor** [comment]
Worked perfectly when running just a template with CMD. But it messed up if ARG is in there.

**Comment by tschettervictor** [comment]
Try this on line 284 this has now worked for me when creating jails and applying individual templates.

args=$(echo "${_line}" | awk -F '[ ]' '{$1=""; sub(/^ */, ""); print $0;}' | eval "sed ${ARG_REPLACEMENTS}")

**Comment by michael-o** [comment]
Suprisingly:

```
awk -F '[ ]' '{$1=""; sub(/^ */, ""); print $0;}'
```
Does the job. The old template as well as the fixed template do work now and `ssh_config` is correct:
```patch
# svn diff -x -u0 .
Index: base-config/Bastillefile
===================================================================
--- base-config/Bastillefile    (Revision 1864)
+++ base-config/Bastillefile    (Arbeitskopie)
@@ -29 +29 @@
-CMD sed -i "" -e 's/# Host \*/Host \*/' -e 's/#\ \ \ GSSAPIAuthentication no/\ \ GSSAPIAuthentication yes/' /etc/ssh/ssh_config
+CMD sed -i "" -e 's/# Host \*/Host \*/' -e 's/#   GSSAPIAuthentication no/  GSSAPIAuthentication yes/' /etc/ssh/ssh_config
```

**Comment by tschettervictor** [comment]
#787

**Comment by tschettervictor** [comment]
Merged.

**Comment by michael-o** [comment]
Hooray!
